<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="MendelGeneDropping.jl : This package performs gene dropping with several useful options for the output.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>MendelGeneDropping.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelGeneDropping.jl">View on GitHub</a>

          <h1 id="project_title">MendelGeneDropping.jl</h1>
          <h2 id="project_tagline">This package performs gene dropping with several useful options for the output.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelGeneDropping.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelGeneDropping.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p>Mendel Gene Dropping is a component of the umbrella <a href="https://openmendel.github.io">OpenMendel</a> project. This analysis option “drops” genes into a pedigree through the founders and propagates them through the pedigrees. In many circumstances, it is useful to simulate genetic data consistent with a postulated map. For example, you might want to generate p-values empirically or to estimate the power of a collection of pedigrees to detect linkage. Gene dropping randomly fills in genotypes subject to prescribed allele frequencies, a given genetic map, and Hardy-Weinberg and linkage equilibrium. After performing gene dropping the package generates a new pedigree file for further analysis. The missing data pattern in the new pedigrees can mimic the input pedigrees, or the user can specify a new pattern.</p>

<h3>
<a id="appropriate-problems-and-data-sets" class="anchor" href="#appropriate-problems-and-data-sets" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Appropriate Problems and Data Sets</h3>

<p>The raw material for gene dropping consists of sets of pedigrees and loci. People within the pedigrees must be assigned either blank phenotypes or Mendelian consistent phenotypes. Gene dropping is carried out independently of observed phenotypes at those loci common to the definition and map files. By varying the content of the map file, you can choose exactly which loci to subject to gene dropping. Phenotypes at the remaining loci of the definition file are left untouched. Simulated genotypes rather than simulated phenotypes are reported. There are no limits on the complexity of the pedigrees or the number of loci. You can use founders from different populations, provided these populations are defined. Each founder should be assigned to a population; any unassigned founders are assumed to come from the first population.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p><em>Note: Three OpenMendel packages - <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays</a>, <a href="https://github.com/OpenMendel/Search.jl">Search</a>, and <a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> must be installed before any Mendel analysis packages will run.</em></p>

<p>Within Julia, use the package manager to install Mendel     :</p>

<pre><code>Pkg.clone("https://github.com/OpenMendel/MendelGeneDropping.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>The MendelGeneDropping analysis package uses the following input files. Example input files can be found in the <a href="https://github.com/OpenMendel/MendelGeneDropping.jl/tree/master/docs">docs</a> subfolder of the MendelGeneDropping project. (An analysis won't always need every file type below.)</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a list of common keywords, see <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">Keywords Table</a>).</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#locus-file">Locus File</a>: Names and describes the genetic loci in your data.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, sex, family structure, and ancestry.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#phenotype-file">Phenotype File</a>: Lists the available phenotypes.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, allele names, allele frequencies.</li>
<li>
<a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set. Must be a standard binary PLINK BED file in SNP major format. If you have a SNP data file you must have a SNP definition file.</li>
</ul>

<h3>
<a id="control-file" class="anchor" href="#control-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control file<a id="control-file"></a>
</h3>

<p>The Control file is a text file consisting of keywords and their assigned values. The format of the Control file is:</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>Below is an example of a simple Control file to run Gene Dropping:</p>

<pre><code>#
# Input and Output files.
#
locus_file = genedropping LocusFrame.txt
pedigree_file = genedropping PedigreeFrame.txt
phenotype_file = genedropping PhenotypeFrame.txt
new_pedigree_file = genedropping NewPedigreeFrame.txt
output_file = genedropping Output.txt
#
# Analysis parameters for GeneDropping option.
#
repetitions = 2
populations = European, African, Chinese
</code></pre>

<p>In the example above, there are seven keywords. Three keywords specify the input files: <em>genedropping LocusFrame.txt</em>, <em>genedropping PedigreeFrame.txt</em>, and <em>genedropping PhenotypeFrame.txt</em>. Two keywords specify the output files: <em>genedropping Output.txt</em> the results file - and <em>genedropping NewPedigreeFrame.txt</em> - the new pedigree OpenMendel generates, with the added simulated genotypes. The last two keywords specify analysis parameters: * repetitions* (2), and <em>populations</em> (European, African, and Chinese). The text after the '=' are the keyword values.</p>

<h3>
<a id="keywords" class="anchor" href="#keywords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keywords<a id="keywords-table"></a>
</h3>

<p>This is a list of OpenMendel keywords specific to Gene Dropping. A list of OpenMendel keywords common to most analysis package can be found <a href="https://openmendel.github.io/MendelBase.jl/#keywords-table">here</a>. The names of keywords are <em>not</em> case sensitive. (The keyword values <em>may</em> be case sensitive.)</p>

<table>
<thead>
<tr>
<th>Keyword</th>
<th>Default Value</th>
<th>Allowed Values</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>gene_drop_output</td>
<td>Unordered</td>
<td>"Unordered", "Ordered", "Sourced", or "Population"</td>
<td>Output format style for gene drop pedigrees</td>
</tr>
<tr>
<td>interleaved</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Whether or not genes are dropped into pedigrees sequentially or interleaved</td>
</tr>
<tr>
<td>keep_founder_genotypes</td>
<td>TRUE</td>
<td>TRUE, FALSE</td>
<td>Whether or not founder genotypes are retained</td>
</tr>
<tr>
<td>missing_rate</td>
<td>0.0</td>
<td>Real</td>
<td></td>
</tr>
<tr>
<td>repetitions</td>
<td>1</td>
<td>integer</td>
<td>Repetitions for sharing statistics</td>
</tr>
</tbody>
</table>

<h3>
<a id="data-files" class="anchor" href="#data-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data Files</h3>

<p>Gene Dropping requires a <a href="https://openmendel.github.io/MendelBase.jl/#control-file">Control file</a>, and a <a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">Pedigree file</a>. Genotype data can be included in the Pedigree file, in which case a <a href="https://openmendel.github.io/MendelBase.jl/#locus-file">Locus file</a> is required. Alternatively, genotype data can be provided in a <a href="https://openmendel.github.io/MendelBase.jl/#snp-data-file">SNP data file</a>, in which case a <a href="https://openmendel.github.io/MendelBase.jl/#snp-definition-file">SNP Definition File</a> is required. OpenMendel will also accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> FAM and BIM files. Details on the format and contents of the Control and data files can be found on the <a href="https://openmendel.github.io/MendelBase.jl">MendelBase</a> documentation page. There are example data files in the Gene Dropping <a href="https://github.com/OpenMendel/MendelGeneDropping.jl/tree/master/docs">docs</a> folder.</p>

<h3>
<a id="running-the-analysis" class="anchor" href="#running-the-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running the Analysis</h3>

<p>To run this analysis package, first launch Julia. Then load the package with the command:</p>

<pre><code> julia&gt; using MendelGeneDropping
</code></pre>

<p>Next, if necessary, change to the directory containing your files, for example,</p>

<pre><code> julia&gt; cd("~/path/to/data/files/")
</code></pre>

<p>Finally, to run the analysis using the parameters in the control file Control_file.txt use the command:</p>

<pre><code> julia&gt; GeneDropping("Control_file.txt")
</code></pre>

<p><em>Note: The package is called</em> MendelGeneDropping <em>but the analysis function is called simply</em> GeneDropping.</p>

<h3>
<a id="interpreting-the-results" class="anchor" href="#interpreting-the-results" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Interpreting the results</h3>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">MendelGeneDropping.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
